### PR TITLE
Implement request timeouts

### PR DESCRIPTION
**Pull request by AndrejMitrovic**
The second commit is me testing stuff to see if it works.

### CODE REVIEW

File: source/geod24/LocalRest.d [hunk 1]
```diff
@@ -87,6 +86,8 @@ private struct Response
     /// `true` if the method returned successfully,
     /// `false` if an `Exception` occured
     bool success;
+    /// `true` if the request timed-out
+    bool timed_out;
```
Comment (Geod24) [inline source/geod24/LocalRest.d]: I'd prefer making `success` an `enum` at this point

Comment (AndrejMitrovic) [reply source/geod24/LocalRest.d]: Good point.

File: source/geod24/LocalRest.d [hunk 2]
```diff
@@ -628,18 +648,50 @@ public final class RemoteAPI (API) : API
...
+
+                            if (this.timeout == Duration.init)
+                            {
+                                return C.receiveOnly!(Response);
+                            }
+                            else
+                            {
+                                bool received_value;
+                                Response actual_res;
+                                C.receiveTimeout(this.timeout,
```
Comment (Geod24) [inline source/geod24/LocalRest.d]: This returns a `bool` which tells you if it timed out

File: source/geod24/LocalRest.d [hunk 3]
```diff
@@ -628,18 +648,50 @@ public final class RemoteAPI (API) : API
...
+                                    {
+                                        received_value = true;
+                                        actual_res = res;
+                                    });
+
+                                if (received_value)
+                                {
+                                    return actual_res;
+                                }
+                                else
```
Comment (Geod24) [inline source/geod24/LocalRest.d]: Technically you don't need to branch here (`return` makes `else` implicit)

File: source/geod24/LocalRest.d [hunk 4]
```diff
@@ -276,16 +288,18 @@ public final class RemoteAPI (API) : API
...
+          timeout = (optional) timeout to use with requests
 
         Returns:
           A `RemoteAPI` owning the node reference
 
     ***************************************************************************/
 
-    public static RemoteAPI!(API) spawn (Impl) (CtorParams!Impl args)
+    public static RemoteAPI!(API) spawn (Impl) (CtorParams!Impl args,
+        Duration timeout = Duration.init)
```
Comment (Geod24) [inline source/geod24/LocalRest.d:304]: I worship whoever implemented default parameter after variadic args

File: source/geod24/LocalRest.d [hunk 5]
```diff
@@ -1176,3 +1228,80 @@ unittest
...
+    import core.thread;
+    import std.exception;
+
+    static interface API
+    {
+        bool checkTimedOut ();
+        size_t sleepFor (long dur);
+    }
+
+    __gshared API api;  // can't pass it through ctor
```
Comment (Geod24) [inline source/geod24/LocalRest.d]: And you don't want to. If you want to contact an API, you store the `Tid` and instantiate a new `RemoteNode`.
`grep` for `tbn` to see how it's done.

Comment (AndrejMitrovic) [reply source/geod24/LocalRest.d]: Wait I fixed this.. where did it go.

**Comment by AndrejMitrovic** [comment]
I pushed some fixes. And now the unittests succeed (locally), but I don't trust it.

**Comment by AndrejMitrovic** [comment]
I mean I am not 99% sure if this is correct code. It makes sense to me, but you're more familiar with the code.

File: source/geod24/LocalRest.d [hunk 6]
```diff
@@ -148,7 +160,20 @@ private final class LocalScheduler : C.FiberScheduler
...
+            if (!ptr.c.wait(duration))
+            {
+                this.pending.status = Status.Timeout;
+                this.pending.data = "";
+            }
+        }
+        else
+        {
+            ptr.c.wait();
+        }
```
Comment (Geod24) [inline source/geod24/LocalRest.d]:
```d
if (duration == Duration.init)
    ptr.c.wait();
else if (!ptr.c.wait(duration))
    this.pending = Response(Status.Timeout, this.pending.id);
```

File: source/geod24/LocalRest.d [hunk 7]
```diff
@@ -1176,3 +1233,77 @@ unittest
...
+    {
+        void check ();
+        size_t sleepFor (long dur);
+    }
+
+    static class Node : API
+    {
+        override void check ()
+        {
+            auto node = RemoteAPI!API.spawn!Node(50.msecs);
```
Comment (Geod24) [inline source/geod24/LocalRest.d]: I would not spawn a `Node` from another thread. At the moment `Node`s should be spawned from the main thread.

**Comment by AndrejMitrovic** [comment]
I pushed a separate commit which sometimes gives me a Json error, but it looks like it's a timing bug..

**Comment by AndrejMitrovic** [comment]
```
~/dev/localrest request-timeouts $ test
Deserializing composite type ArgWrapper!() which has no serializable fields.
Serializing composite type ArgWrapper!() which has no serializable fields
All unit tests have been run successfully.

~/dev/localrest request-timeouts $ test
Deserializing composite type ArgWrapper!() which has no serializable fields.
Serializing composite type ArgWrapper!() which has no serializable fields
Program exited with code -11
```

And sometimes a segfault.

File: source/geod24/LocalRest.d [hunk 8]
```diff
@@ -1176,3 +1233,84 @@ unittest
...
+            return 42;
+        }
+    }
+
+    auto node_1 = RemoteAPI!API.spawn!Node();
+    auto node_2 = RemoteAPI!API.spawn!Node();
+    node_tid = node_2.tid;
+    node_1.check();
+}
+
```
Comment (Geod24) [inline source/geod24/LocalRest.d]: That's probably the root of all your problems. How could you not see this ?

**Comment by AndrejMitrovic** [comment]
The issue was the timeout was set too low, so it was an earlier assert that failed (which did not expect a timeout to occur).

I've changed the timeout to 500 msecs instead of 50 msecs. It went green once, let's see if it goes green again.

**Comment by AndrejMitrovic** [comment]
@Geod24 green twice in a row. :)

**Pull request merged**